ext/standard/var.c

```diff
@@ -782,13 +782,20 @@ static int php_var_serialize_try_add_sleep_prop(
 	if (Z_TYPE_P(val) == IS_INDIRECT) {
 		val = Z_INDIRECT_P(val);
 		if (Z_TYPE_P(val) == IS_UNDEF) {
-			zend_property_info *info = zend_get_typed_property_info_for_slot(Z_OBJ_P(struc), val);
-			if (info) {
-				zend_throw_error(NULL,
-					"Typed property %s::$%s must not be accessed before initialization (in __sleep)",
-					ZSTR_VAL(Z_OBJCE_P(struc)->name), ZSTR_VAL(error_name));
+			zend_property_info *prop_info = zend_get_typed_property_info_for_slot(Z_OBJ_P(struc), val);
+			if (!prop_info) {
+				return FAILURE;
 			}
-			return FAILURE;
+			zval *prop;
+			if ((prop_info->flags & ZEND_ACC_STATIC) != 0) {
+				prop = &prop_info->ce->default_static_members_table[prop_info->offset];
+			} else {
+				prop = &prop_info->ce->default_properties_table[OBJ_PROP_TO_NUM(prop_info->offset)];
+			}
+			if (Z_ISUNDEF_P(prop) || !ZEND_TYPE_ALLOW_NULL(prop_info->type)) {
```

Comment:
Should probably document that in the case that ZEND_TYPE_ALLOW_NULL(prop_info->type), the serializer will later convert IS_UNDEF to null in the serialized value as the closest available equivalent.